Runtime dispatch based on custom CPU capabilities function

dkostic opened the pull request:
Addresses #1148, and as agreed therein, the dispatch logic is implemented in the backends.

mkannwischer reviewed (changes requested):
Thanks @dkostic! This looks excellent. Some minor comments/questions.

Before we merge this, we should definitely add a test configuration that hardcodes `mlk_is_native_capable` to zero to test this in CI.
There are already some custom configuration tests [here](https://github.com/pq-code-package/mlkem-native/blob/51d641ad07f0df8b61a2aafbb553d67d1523e685/.github/workflows/ci.yml#L451), so this should be easy. I can help with adding that if you want.

mkannwischer commented:
CBMC proofs appear to blow up. That's something that also needs to be fixed before we can merge. 
@rod-chapman, could you maybe help with that?

hanno-becker commented on mlkem/src/native/x86_64/meta.h:
```diff
@@ -43,131 +44,275 @@ static MLK_INLINE int mlk_rej_uniform_native(int16_t *r, unsigned len,
+  else
+  {
+    mlk_ntt_avx2((__m256i *)data, mlk_qdata.vec);
+    return 1;
```
The toplevel API uses `0` as the return code for success, which I'd suggest to stick with.

dkostic replied:
changed to -1 and 0 for fail/success.

hanno-becker reviewed (changes requested):
Can we keep the original C functions as `foo_c()` and then have something like
```
void foo() {
   if (foo_native() == 0) {
      return;
   }
   foo_c();
}
```
for the dispatch? That modularizes the reasoning.

hanno-becker commented:
On the other hand, having a dedicated function for the dispatch creates a lot of churn for the CBMC proofs because a lot needs renaming, and additional proofs be created for wrappers. I am undecided.

Why are the CBMC proofs currently timing out? Can this be investigated first?

mkannwischer commented:
> 3. foo_c() and foo_native() should have _identical_ CBMC contracts, except that foo_native can return "0" meaning "success" or "1" meaning "can't do that... please use the C back-end instead".

We should stick to `-1` for the latter, because rejection sampling will return >= 0 in case of success.

> 
> Does that make sense?

@rod-chapman, @hanno-becker, can you elaborate a bit on what the benefit would be of splitting of the C part into it's own function? I see no immediate benefit of doing that and it is _a lot_ of work. For CBMC, it should look exactly the same? The resulting binary will look the same too.

rod-chapman commented:
I know what the problem is - at least with poly_ntt(). In short - when the call to poly_ntt_native() _fails_, we need to know that the polynomial p[] _hasn't_ changed at all, so that the subsequent loop invariant and the pre-condition on the call to ntt_layer() can be satisfied.  In short - poly_ntt_native() needs a stronger post-condition. Working on it now.

dkostic commented:
Everything is passing now except this one that keeps saying "download buffer is full": https://github.com/pq-code-package/mlkem-native/actions/runs/16975743408/job/48128706572?pr=1152. I tried restarting it twice, but no success so far.

dkostic commented:
@mkannwischer @hanno-becker @rod-chapman the CI is passing now and I think I addressed all your comments. Let me know if you have any additional questions/concerns!

hanno-becker reviewed (changes requested):
* I believe the CBMC proofs for `xxx_native()` must work with the contract for `mlk_is_native_capable`, otherwise the hardcoded default of `1` will render the `FAIL` branch unreachable.
* If possible, I would like to see the native API strengthened to include that for failure, the output buffer is unchanged. I think @rod-chapman you have experimented with this already?

Also, can we make the name of the native capability function more specific? I think there should be one backend capability function per feature (set) that the backend relies on? Starting with AVX2, we could have `mlk_avx2_capable()` or similar.

dkostic commented:
@hanno-becker I tried to move the capability function to `sys.h` and tried different combinations of flags to make it work with all build flavors in `examples`, but I can't make it work. I think the main issue is this: https://github.com/pq-code-package/mlkem-native/blob/51d641ad07f0df8b61a2aafbb553d67d1523e685/scripts/autogen#L1260. Let me know if you have an idea how to make it work, otherwise, I'll revert back to a separate file.

dkostic commented:
> Also, can we make the name of the native capability function more specific? Should there not be one capability function per feature (set) that the backend relies on? E.g. for AVX2, we could have mlk_avx2_capable() or similar.

From our previous chats I understood that you don't want that level of detail in mlkem-native, but I can do it nevertheless. But let's see first if it's possible to have the functions in `sys.h` because that would simplify the implementation (otherwise, I would have to add a file per backend for capability functions).

hanno-becker commented:
@dkostic I think the issue is fixed.

@rod-chapman Thanks for strengthening the specs. I added a macro to encapsulate the 'array unchanged' logic.
